### PR TITLE
[TOSA] Add Torch to Tosa Legalization for torch.tril

### PR DESCRIPTION
**Pull request by justin-ngo-arm**
Change-Id: Ie5ba31a27394c3adcea00266a9d562862dbd8b08

### CODE REVIEW

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -5389,6 +5390,112 @@ ConvertAtenOp<Aten__InterpolateSizeListScaleListOp>::matchAndRewrite(
   return success();
 }
 
+// Template to create support tril support const tensor for aten.tril
```
Comment (mgehre-amd) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:5393]: nit: wording

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -5389,6 +5390,112 @@ ConvertAtenOp<Aten__InterpolateSizeListScaleListOp>::matchAndRewrite(
...
+// legalization
+template <typename T>
+Value createTrilConstTensor(PatternRewriter &rewriter, Operation *op,
```
Comment (mgehre-amd) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:5396]:
```suggestion
Value createTrilMask(PatternRewriter &rewriter, Operation *op,
```
maybe?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 3]
```diff
@@ -5389,6 +5390,112 @@ ConvertAtenOp<Aten__InterpolateSizeListScaleListOp>::matchAndRewrite(
...
+  if (!selfType)
+    return rewriter.notifyMatchFailure(
+        op, "Only ranked tensor types are supported");
+
+  // Rank below 2 not accepted
+  auto selfRank = selfType.getRank();
+  if (selfRank <= 1)
+    return rewriter.notifyMatchFailure(
+        op, "Rank 0 and 1 are not accepted as they cause underflow");
+
```
Comment (mgehre-amd) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:5461]: Can you please check that the element type is one of the ones accepted by getTrilConstTensor?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 4]
```diff
@@ -5389,6 +5390,112 @@ ConvertAtenOp<Aten__InterpolateSizeListScaleListOp>::matchAndRewrite(
...
+  SmallVector<int64_t> constShape;
+  for (auto i = 0; i < selfRank - 2; i++)
+    constShape.push_back(1);
+  constShape.push_back(h);
+  constShape.push_back(w);
+
+  const TypeConverter *typeConverter = this->getTypeConverter();
+  RankedTensorType resultType = cast<RankedTensorType>(
+      typeConverter->convertType(op->getResult(0).getType()));
+
```
Comment (mgehre-amd) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:5489]: can we check that `resultType` is not empty?

**Review by mgehre-amd — APPROVED** [review]

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (justin-ngo-arm) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:5393]: Ack

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 2 above)
Comment (justin-ngo-arm) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:5396]: Ack, that does sound better.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 4 above)
Comment (justin-ngo-arm) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:5489]: Ack

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 3 above)
Comment (justin-ngo-arm) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:5461]: I have checked and tested passing the element type to getTrilConstTensor, and it's been working correctly so far.

**Pull request merged**